BASE/src/Microsoft.ApplicationInsights/Channel/RedirectHttpHandler.cs

```diff
@@ -0,0 +1,130 @@
+﻿namespace Microsoft.ApplicationInsights.Channel
+{
+    using System;
+    using System.Net;
+    using System.Net.Http;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    using Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing;
+
+    internal class RedirectHttpHandler : HttpClientHandler
+    {
+        internal const int MaxRedirect = 10;
+
+        private readonly Cache<Uri> cache = new Cache<Uri>();
+
+        public RedirectHttpHandler()
+        {
+            this.AllowAutoRedirect = false;
+        }
+
+        /// <summary>
+        /// This method will handle all requests from HttpClient.SendAsync().
+        /// This inspects <see cref="HttpResponseMessage"/> to see if we received a redirect from the Ingestion service.
+        /// </summary>
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            if (this.cache.TryRead(out Uri redirectUri))
+            { 
+                request.RequestUri = redirectUri;
+            }
+
+            HttpResponseMessage response = await base.SendAsync(request, cancellationToken).ConfigureAwait(false);
+
+            if (IsRedirection(response.StatusCode))
+            {
+                return await this.HandleRedirectAsync(request, response, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                return response;
+            }
+        }
+
+        private static bool IsRedirection(HttpStatusCode statusCode)
+        {
+            switch ((int)statusCode)
+            {
+                case 307: // StatusCodes.Status307TemporaryRedirect
+                case 308: // StatusCodes.Status308PermanentRedirect
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
+        private static bool TryGetRedirectVars(HttpResponseMessage httpResponseMessage, out Uri redirectUri, out TimeSpan expire)
+        {
+            expire = httpResponseMessage?.Headers?.CacheControl?.MaxAge ?? default;
+            redirectUri = httpResponseMessage?.Headers?.Location;
+
+            return expire != default && redirectUri != null && redirectUri.IsAbsoluteUri;
```

Comment:
I just want to point out that setting the CacheControl on a POST is... ambiguous at best. I recognize it's what Breeze does, but I think it's non-standard. 

My understanding is that cache-control relates to the content returned, not to the Location header.
Of course, Breeze is free to use cache-control this way, if they want.

The upshot is that this class is tuned to work only with Breeze -- as long as they continue to set cache-control on POST. Specifically, though, if any other endpoint returns 308, setting the location header but not cache_control, then we won't actually follow the redirect.

Perhaps you could drop the `expire != default` clause, making the cache-control optional. You'd still follow the redirect, but wouldn't cache the resulting